controllers/layered_image_handler.go

```diff
@@ -0,0 +1,149 @@
+		if err != nil {
+			r.Log.Info("Error in retrieving LayeredImageDeployment ConfigMap", "err", err)
+			return err
+		}
+
+		ImgMc, err = r.createMachineConfigFromConfigMap(cm)
+		if err != nil {
+			r.Log.Info("Error in creating MachineConfig for LayeredImageDeployment from ConfigMap", "err", err)
+			return err
+		}
+
+	} else {
+		r.Log.Info("LayeredImageDeployment feature is disabled. Resetting ImgMc")
+		// Reset ImgMc
+		ImgMc = nil
+
+	}
+
+	return nil
+}
+
+func (r *KataConfigOpenShiftReconciler) getExistingMachineConfig() (*mcfgv1.MachineConfig, error) {
+	r.Log.Info("Getting any existing MachineConfigs related to KataConfig")
+
+	// Retrieve the existing MachineConfig
+	// Check for label "app":r.kataConfig.Name
+	// and name "50-enable-sandboxed-containers-extension" or name "50-enable-sandboxed-containers-image"
+	mcList := &mcfgv1.MachineConfigList{}
+	err := r.Client.List(context.Background(), mcList)
+	if err != nil {
+		r.Log.Info("Error in listing MachineConfigs", "err", err)
+		return nil, err
+	}
+
+	for _, mc := range mcList.Items {
+		if mc.Labels["app"] == r.kataConfig.Name &&
+			(mc.Name == extension_mc_name || mc.Name == image_mc_name) {
+			return &mc, nil
+		}
+	}
+
+	r.Log.Info("No existing MachineConfigs related to KataConfig found")
+
+	return nil, nil
+}
+
+// Method to create a new MachineConfig object from configMap data
+// The configMap data will have two keys: "osImageURL" and "kernelArgs"
+func (r *KataConfigOpenShiftReconciler) createMachineConfigFromConfigMap(cm *corev1.ConfigMap) (*mcfgv1.MachineConfig, error) {
+
+	// Get the osImageURL from the ConfigMap
+	// osImageURL is mandatory for creating a MachineConfig
+	osImageURL, exists := cm.Data["osImageURL"]
+	if !exists {
+		return nil, fmt.Errorf("osImageURL not found in ConfigMap")
+	}
+
+	ic := ignTypes.Config{
+		Ignition: ignTypes.Ignition{
+			Version: "3.2.0",
+		},
+	}
+
+	icb, err := json.Marshal(ic)
+	if err != nil {
+		return nil, err
+	}
+	mc := &mcfgv1.MachineConfig{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "machineconfiguration.openshift.io/v1",
+			Kind:       "MachineConfig",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      image_mc_name,
+			Namespace: OperatorNamespace,
+		},
+		Spec: mcfgv1.MachineConfigSpec{
+			Config: runtime.RawExtension{
+				Raw: icb,
+			},
+			OSImageURL: osImageURL,
+		},
+	}
+
+	if kernelArguments, ok := cm.Data["kernelArguments"]; ok {
+		// Parse the kernel arguments and set them in the MachineConfig
+		// Note that in the configmap the kernel arguments are stored as a single string
+		// eg. "a=b c=d ..." and we need to split them into individual arguments
+		// eg ["a=b", "c=d", ...]
+		// Split the kernel arguments string into individual arguments
+		mc.Spec.KernelArguments = strings.Fields(kernelArguments)
+	}
+
+	return mc, nil
+}
+
+// Method to retrieve Image MachineConfig from the cluster
+func (r *KataConfigOpenShiftReconciler) getImageMc() *mcfgv1.MachineConfig {
+	return ImgMc
```

Comment:
Is it possible to generate the `MachineConfig` when `getImageMc` is called?

Reply:
Yeah, it is possible. 
I can check if the featuregate is enabled and generate the MachineConfig in this method.
However, since the approach I took is to do all these in the handler function, I avoided generating it again in this function.

Reply:
I find it unusual, and probably confusing, to define `ImgMc` before it's needed, and to keep it in memory until it's applied.